### DISCUSSION TITLE
some values assigned incorrectly, some correctly.

### DISCUSSION THREAD

**Discussion by mophilly** [enoch3712/ExtractThinker#127] [Q&A] [answered]
I would like to better understand how to guide the extractor and or parser to assign values to the correct attribute. Links to articles or documents in the project would be great.

In the example model below, I have used the label names on the PDF as the field names in the model, but concessions made for clarity of purpose.

This model is working for the most part. There are two cases where values don't end up in the proper place. One is class Transaction and the other class StatementTotal. 

Note the values of owner_deductions and owner_taxes in the expected result and the actual result. class Transaction has a similar problem where values end up in the wrong column and some values duplicated to columns that should empty or None. The source PDF presents only white space where a value might be when there is no value to report; as compared to placing a zero in that spot.

Expected output, as json, for StatementTotal, drawn from the source PDF:
````
"check_totals": {
		"payee_gross_value": 146.11,
		"payee_deductions": null,
		"payee_taxes": -6.99,
		"other_deductions": null,
		"payee_net_value": 139.12
	},
	"year_to_date": {
		"payee_gross_value": 146.11,
		"payee_deductions": 6.99,
		"payee_taxes": null,
		"other_deductions": null,
		"payee_net_value": 139.12
	}
````

Actual result, as json
````
"check_totals": {
		"payee_gross_value": null,
		"payee_deductions": -6.99,
		"payee_taxes": null,
		"other_deductions": null,
		"payee_net_value": 139.12
	},
	"year_to_date": {
		"payee_gross_value": null,
		"payee_deductions": -6.99,
		"payee_taxes": null,
		"other_deductions": null,
		"payee_net_value": 139.12
	}
````

example pydantic model:
```
Class MonthlyStatement
	Header: Class Header
		Payee: class Company
		Payor: class Company
		Check info: class Check_Info
	Property: List[class Property]
		Property ID
		Property Name
		Products: List[class Product]
			Product code
			Transactions details: List[class Transaction]
				Transaction code
				Transaction_details: List[class TransactionDetail]
					td_code
					td_description
					td_quantity
					td_amount
			Transaction Total
		Property total: class PropertyTotal
			Property Total
	Payee total: class StatementTotal
	Year To Date: class StatementTotal
				
class ReportTotal(Contract):
    payee_gross_value: Optional[float] = None
    payee_deductions: Optional[float] = None
    payee_taxes: Optional[float] = None
    other_deductions: Optional[float] = None
    payee_net_value: Optional[float] = None

```

**Comment by enoch3712** [comment, marked as answer]
Sorry for the delay @mophilly 

Just published the new version, please update to 0.0.28

On this point, is really simple, just remove the optionals and a description for good measure:

```
class ReportTotal(Contract):
    payee_gross_value: float = Field(description="The gross value of the payee, column name: 'Gross Value' position: 0")
    payee_deductions: float = Field(description="The deductions of the payee, column name: 'Deductions' position: 1")
    payee_taxes: float = Field(description="The taxes of the payee, column name: 'Taxes' position: 2")
    other_deductions: float = Field(description="The other deductions of the payee, column name: 'Other Deductions' position: 3")
    payee_net_value: float = Field(description="The net value of the payee, column name: 'Net Value' position: 4")
```

This will most likely fix your problem. 

> I would like to better understand how to guide the extractor and or parser to assign values to the correct attribute. Links to articles or documents in the project would be great.

The primary parsing is done by instructor, the most popular parsing library, that basically uses reflection and other patterns to enforce structure. 

I go into detail here on the extraction section: https://levelup.gitconnected.com/claude-3-5-the-king-of-document-intelligence-f57bea1d209d

instructor - https://github.com/instructor-ai/instructor

Also pydanticai, that does the same but is done by the Pydantic team. https://ai.pydantic.dev/

Pydantic will be added shortly into the code, so you will be able to pick an "llm agent" to make the extraction

**Reply by mophilly** [reply]
Version 0.0.29 seems to have improved the output, but the output is not yet correct. I like the Field description; it improves the code a lot. I am reading [https://docs.pydantic.dev/2.10/concepts/fields/](https://docs.pydantic.dev/2.10/concepts/fields/).

In the source document the value for property_qty is a blank space, but the pydantic class definition calls it a float. I haven't included a default value. I plan to try that, e.g. ```property_qty float = Field(description="units sold", default=0)```

Note that the values are not in the desired places. Also, the value '0.11' appears twice when it should be only assigned to "owner_share".

Here is what is expected:
```{
	"owner_code": "10.RI",
	"owner_code_type": "REVENUE INTEREST",
	"production_date": "Nov 22",
	"property_qty":0,
	"property_gross": 8983.30,
	"property_price": 77.28,
	"property_value": 694184.51,
	"owner_interest": 0.00001254,
	"owner_distribution_interest": 0.00001254,
	"owner_share": 0.11,
	"owner_value": 8.71
}
```

Here is what I got:
```{
	"owner_code": "10.RI",
	"owner_code_type": "REVENUE INTEREST",
	"production_date": "Nov 22",
	"property_qty": 77.28,
	"property_gross": 694184.51,
	"property_price": 1.254e-05,
	"property_value": 8.71,
	"owner_interest": 0.11,
	"owner_distribution_interest": 1.254e-05,
	"owner_share": 0.11,
	"owner_value": 8.71
}
```

**Reply by enoch3712** [reply]
@mophilly what models are you using? 

This should working just fine with vision. Did you try extraction with vision=true? 

Most of this problem seems to be that you are using a documentLoader that makes it difficult for the model to get the data in the proper places

**Reply by mophilly** [reply]
I am using get-40-mini. I am happy to switch if there is a better choice because your project makes that pretty easy!

I pass vision=true but I don't really know what that does or how to check that it is working on my set up.

Here is my invocation:
```# create a document loader
    document_loader = DocumentLoaderPyPdf()

    # create the extractor
    extractor = Extractor()
    extractor.load_document_loader(document_loader)
    extractor.load_llm("gpt-4o-mini")  # or any other supported model

    # Extract data from your document
    result = extractor.extract(input_path, RunStatementContract, vision=True)
```

**Reply by enoch3712** [reply]
Vision generates an image of each page, and injects it together with the content from the documentLoader. 

So, will fix this problems you should have. **"gpt-4o-mini"** should work fine for most of this use cases. Just for good measure try **"gpt-4o"** if makes sense

**Reply by mophilly** [reply]
I changed the pydantic class definitions, see #138 for details, and I switched to gpt-4o. Now the result is aligned with the source. 

Thanks for all the advice!
